lib/archive/archive.go

```diff
@@ -109,13 +109,31 @@ func Extract(r io.Reader, dir string) error {
 			return trace.Wrap(err)
 		}
 
-		if err := extractFile(tarball, header, dir); err != nil {
+		if err := extractFile(tarball, header, dir, header.Name); err != nil {
 			return trace.Wrap(err)
 		}
 	}
 	return nil
 }
 
+// ExtractGlob extracts the contents of the specified tarball under directory dir.
+// Only the files/directories found in tarDir in the tarball matching patterns are extracted.
+// Note, files from tarDir are written directly to dir omitting tarDir.
+// The resulting files and directories are created using the current user context.
+func ExtractGlob(r io.Reader, dir, tarDir string, patterns []string) error {
+	return trace.Wrap(TarGlob(tar.NewReader(r), tarDir, patterns, func(tarRelPath string, header *tar.Header, f *tar.Reader) error {
+		// Security: ensure tar doesn't refer to file paths outside the directory
+		err := SanitizeTarPath(header, dir)
```

Comment:
Nit: I'd try to be consistent with how err is handled at least in the same function, i.e. let's use either

```go
if err := xxx(); err != nil {
```

style or

```go
err := xxx()
if err != nil {
```

in both places.